### PR TITLE
[recipe] feat: Add DeepSeek-V4-Flash pretraining recipes

### PR DESCRIPTION
**Pull request by weijiac0619**
# What does this PR do?

Adds pretraining recipes for DeepSeek-V4-Flash on Blackwell, plus a Slurm
launcher example, unit tests, and a functional test. The base recipe targets
TP=1 / PP=4 / EP=8 with selective activation recompute, an MTP-aware pipeline
layout, and BF16. Two variants extend the base for the two supported
optimizer + precision combinations.

# Changelog

- `src/megatron/bridge/recipes/deepseek/deepseek_v4.py` (new): three pretrain
  configs and a pipeline-layout helper.
  - `deepseek_v4_flash_pretrain_config()` — BF16 base; TP=1, PP=4, EP=8,
    selective recompute (`moe_act`, `mhc`), MTP placed on the last PP stage
    via `pipeline_model_parallel_layout`.
  - `deepseek_v4_flash_pretrain_mxfp8_config()` — Adam + MXFP8 training,
    BF16 MTP / validation eval, `quant_recipe` selects MXFP8 for TE linears.
  - `deepseek_v4_flash_pretrain_muon_config()` — Muon optimizer + BF16,
    non-layer-wise dispatch.
  - `set_deepseek_v4_pipeline_model_parallel_layout()` helper builds the
    even decoder layout with MTP and loss on the last PP rank.
- `src/megatron/bridge/recipes/deepseek/__init__.py`: re-export the new
  configs and helper.
- `examples/models/deepseek_v4/README.md`: document the new recipes and
  Slurm launcher.
- `examples/models/deepseek_v4/slurm_pretrain.sh`: Slurm sbatch script for
  the new recipes.
- `tests/unit_tests/recipes/test_deepseek_recipes.py`: extend coverage to
  DSv4 configs.
- `tests/functional_tests/test_groups/recipes/test_deepseek_recipes_pretrain.py`
  (new): L0/L1 pretrain smoke for DSv4 variants.

# GitHub Actions CI

See the [CI section](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md#-running-github-ci) in the Contributing doc for how to trigger the CI.

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?

### CODE REVIEW

File: examples/models/deepseek_v4/slurm_pretrain.sh [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+CONTAINER_IMAGE="${CONTAINER_IMAGE:-}"
+# CONTAINER_IMAGE="/path/to/container.sqsh"
+
+# Container mounts (optional, comma-separated for srun --container-mounts).
+CONTAINER_MOUNTS="${CONTAINER_MOUNTS:-}"
+# CONTAINER_MOUNTS="/data:/data,/workspace:/workspace"
+
+CASE_NAME="${CASE_NAME:-${RECIPE_NAME}}"
+JOB_ID="${SLURM_JOB_ID:-manual}"
+OUTDIR="${OUTDIR:-${WORKSPACE}/results/${MODEL_NAME}_${CASE_NAME}_${JOB_ID}}"
```
Comment (cuichenx) [inline examples/models/deepseek_v4/slurm_pretrain.sh]: can we simplify these -- only keep the parameters that users would possibly change. Keep the other ones at a default value

File: src/megatron/bridge/recipes/deepseek/deepseek_v4.py [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+from megatron.bridge.recipes.utils.optimizer_utils import (
+    distributed_fused_adam_with_cosine_annealing,
+    distributed_muon_with_cosine_annealing,
+)
+from megatron.bridge.training.comm_overlap import CommOverlapConfig
+from megatron.bridge.training.config import ConfigContainer
+from megatron.bridge.training.mixed_precision import bf16_mixed, bf16_with_mxfp8_mixed
+
+
+DSV4_CSA_BACKEND = Literal["unfused", "cudnn_dsa", "tilelang_official", "flashmla_official"]
```
Comment (cuichenx) [inline src/megatron/bridge/recipes/deepseek/deepseek_v4.py]: I think we can remove tilelang_official and flashmla_official now, they were added while debugging forward pass parity

File: src/megatron/bridge/recipes/deepseek/deepseek_v4.py [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+    cfg = _pretrain_common()
+    cfg.model = AutoBridge.from_hf_pretrained(hf_path, trust_remote_code=True).to_megatron_provider(load_weights=False)
+
+    _set_deepseek_v4_common_model_config(cfg, csa_backend=csa_backend, use_fused_kernels=use_fused_kernels)
+    _set_deepseek_v4_common_training_config(cfg)
+    _set_deepseek_v4_optimizer_and_precision(cfg, optimizer_type=optimizer_type, mxfp8=mxfp8)
+    return cfg
+
+
+def deepseek_v4_flash_pretrain_mxfp8_config(hf_path: str = "deepseek-ai/DeepSeek-V4-Flash") -> ConfigContainer:
```
Comment (yaoyu-33) [inline src/megatron/bridge/recipes/deepseek/deepseek_v4.py]: this seems way too many levels created by agent, just use a single at most 2 levels of config system, flatten as much as you can.

Comment (weijiac0619) [reply src/megatron/bridge/recipes/deepseek/deepseek_v4.py]: resolved

File: examples/models/deepseek_v4/README.md [hunk 4]
```diff
@@ -26,9 +26,21 @@ Use `./scripts/switch_mcore.sh main` and `uv sync --locked` to return to the pin
 
 - `conversion.sh` imports HF weights into Megatron Bridge and exports Megatron checkpoints back to HF format.
 - `inference.sh` runs text generation against an HF or Megatron checkpoint.
+- `slurm_pretrain.sh` runs the DeepSeek-V4-Flash pretraining recipes.
```
Comment (yaoyu-33) [inline examples/models/deepseek_v4/README.md:29]: pin mcore dev commit

Comment (yaoyu-33) [reply examples/models/deepseek_v4/README.md:29]: remark: not work inside nemofw container, need to swtich to ^^ dev commit.

**Review by cuichenx — DISMISSED** [review]

**Comment by cuichenx** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/3893/commits/8ecd0470c4e41bd1dcb389a4e514c08c3993c3fc

**Comment by cuichenx** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/3893/commits/0bfb0eb9a8ef23cc22bea3125fbef751699e01e9

**Review by cuichenx — APPROVED** [review]
